Support group volume changes for fixed-volume players using external volume control

SamInPgh opened the pull request:
Add group volume support for players using the Denon/Marantz AVR Control plugin. (Also, white space cleanup.)

Signed-off-by: Sam Y <syahres@gmail.com>

SamInPgh committed:
Update Plugin.pm

Add group volume support for players using the Denon/Marantz AVR Control plugin. (Also, white space cleanup.)

Signed-off-by: Sam Yahres <syahres@gmail.com>

philippe44 commented:
Thanks but is there a way to make that less model-specific? Is there a more generic attribute to use?

michaelherger commented:
@SamInPgh would `digitalVolumeControl` be invalid for your player types?

SamInPgh commented:
Not that I can think of. That's about as specific as it gets. That particular string is populated in the model name attribute by the Denon/Marantz AVR Control plugin during a player's registration with the plugin and should never be found there otherwise.

By the way, welcome back!

philippe44 commented:
@SamInPgh - Thanks. To @michaelherger's question, is there something that is generic and that would valide the volume memorization even if digital volume is invalid? The idea is that I want to avoid as much as possible to make the plugin dependent of other plugins and even more very specific parts of them. If the name change for any reason, then this modification would stop working.

SamInPgh commented:
> @SamInPgh would `digitalVolumeControl` be invalid for your player types?

'digitalVolumeControl' is set to '0' for these players (volume fixed at 100%), which is why they were formerly excluded from group volume control by the plugin.

SamInPgh commented:
> @SamInPgh - Thanks. To @michaelherger's question, is there something that is generic and that would valide the volume memorization even if digital volume is invalid? The idea is that I want to avoid as much as possible to make the plugin dependent of other plugins and even more very specific parts of them. If the name change for any reason, then this modification would stop working.

I understand your concern but the answer is ''No". However I can promise not to change the string. ☺️ And, even if it did change, it wouldn't hurt anyone except those using the Denon plugin, which would be my problem.

EDIT: Just to clarify my "No" answer, I meant that the "Denon/Marantz AVR" string actually IS what "would validate the volume memorization even if digital volume is invalid" (actually not invalid; just set to "off"). I suppose that we could define another more generic attribute in the client object to indicate this player is controlling a downstream amplifier via a plugin that needs to be notified of volume changes even though its volume control option is set to "fixed at 100%".

michaelherger commented:
I think the idea should be to introduce a new player class method or property telling the Group Player plugin whether to synch volume or not. Groups then could check for the existence of this method and its result.

SamInPgh commented:
See the recent edit of my last comment. (Sorry. I should have created a new one).

SamInPgh commented:
_I suppose that we could define another more generic attribute in the client object to indicate this player is controlling a downstream amplifier via a plugin that needs to be notified of volume changes even though its volume control option is set to "fixed at 100%"._

Apps such as Material Skin and Squeeze Ctrl define an option to handle this situation, e.g. for Squeeze Ctrl:

![Screenshot_20250210-013609](https://github.com/user-attachments/assets/fd41bb41-8983-451b-a748-fc9da334a2be)

So maybe the best way to handle it would be for the Group plugin to add its own analagous option, thus maintaining consistency with the client apps.

SamInPgh commented:
Having slept on it, I am now of the opinion that this indeed should be an option in the Groups plugin, just as it is in the client apps. If there is agreement, I will make the necessary changes.

michaelherger commented:
I personally believe that instead of implementing this a third time there should be one solution for them all. The existing solutions were done lack of a better alternative. When there are three (or more) applications in the need of the same solution, it's time to think bigger, to think about a generic solution available to any client implementation.

Consider this expression of a disagreement 😁.

SamInPgh commented:
I believe that this option exists not only in Material Skin and Squeeze Ctrl, but in all client apps except the original LMS skins. This would include Squeezer, Orange Squeeze (now Open Squeeze) and iPeng. I only mentioned MS and SC because they were handy for me to check.

So the question becomes "Do we want to come up with some sort of universal generic solution that involves changing at least 5 other popular client apps / plugins, or add the same simple option to the Groups plugin that they have been using for years?" IMHO, it's a no-brainer.

SamInPgh commented:
Okay, @philippe44. It's your plugin, and you have the tie-breaking vote. My proposal is to add a config option similar to what the client apps use for now, and begin to think about ways that the underlying situation might be addressed more generically under the covers in future releases of LMS. 

Btw, this thread is reminiscent of the "Incremental volume change on muted players" issue that I advocated for changing in LMS five years ago before losing the argument. I predicted at the time that, if it wasn't changed in LMS, every client app and plugin that wanted to support unmuting via volume change in a logical manner would end up coding their own workaround to the illogical way that LMS did it (and still does). Sure enough, that's exactly what ended up happening, and one of the more recent plugins to work around it was this one --- just last year as a matter of fact. Not saying "I told you so" but... I told you so. 😄

philippe44 commented:
I've not looked at the code for a long while but if you need actual volume command, why setting digital volume?

SamInPgh commented:
It's really quite simple. The players that set the volume at "Fixed at 100%" (not limited to the Denon players btw) are telling LMS not to process volume change commands internally for these players but to pass them along to any listeners who may have registered to receive them. The original LMS client skins do nothing special with volume changes for these fixed volume players, letting LMS handle it as intended. However, at some point, other client apps and plugins began the practice of interpreting this as "ignore volume commands for players that have fixed volume", probably due to a misguided user's complaint or something. Who knows? This change then caused developers of plugins such as Denon AVR Control to complain that they were no longer receiving the volume control notifications that their code relied on. So the client app developers of course decided that, in order to please everyone, they should put an option in their apps for whether or not to ignore volume changes for players with fixed volume. For whatever reason, this practice, based on a misunderstanding, has spread to practically all clients apps and plugins. So the "generic" solution to the problem would be to simply stop the misguided practice of preventing client volume control commands from being sent for players with fixed volume. The Groups plugin could be the first to do so by simply removing the check for 'digitalVolumeControl'. If that step seems too drastic, then just follow the lemmings and add the user option for what to do in this situation. I'll leave the decision and any resulting code changes up to you.

michaelherger commented:
Don't know whether I follow everything. But would it be possible to have a flag to "pass on the volume information anyway", even when processing the volume was disabled?

SamInPgh commented:
Yes. As stated previously, that is currently the standard practice in the client apps and I would be happy to make the changes to implement it here if there is agreement.

SamInPgh committed:
Merge branch 'philippe44:master' into master

SamInPgh committed:
Add plugin option for group volume changes

Optionally allow group volume changes to be sent for players with fixed volume.

SamInPgh commented:
See my latest commit, which replaces the check for "Denon/Marantz AVR" model name with a user config option similar to what the client apps use.

mikeysas commented:
@SamInPgh - with this latest change where you would send volume changes to all players in the group, how would you expect this should work for another player in the group with 100% fixed volume that is not using your plugin and whose volume should not change?

In my "mega" Group Player that we have been doing the previous testing with, the Living Room SBT is the player using your plugin and controlling my Denon AVR.  The WiiM Pro Plus is also set to Fixed 100%, but does not use your plugin and I would not want the volume to change on that.

mikeysas commented:
If I understand @michaelherger suggestion correctly....   I like the idea that generically there is a means for a player to be configured as 100% Fixed with volume changes controlling and external device whether that is managed by your plugin or through a user setting a preference.  Then Material could get rid of its global setting which I have never loved for the same reason stated above.  In Material, my WiiM Pro Plus appears as those the volume can be changed and I would rather it not.  This approach would solve that also.

SamInPgh commented:
> _@SamInPgh - with this latest change where you would send volume changes to all players in the group, how would you expect this should work for another player in the group with 100% fixed volume that is not using your plugin and whose volume should not change?_
>

The same way it works if you turn the corresponding option on in Material Skin, Squeeze Ctrl, Squeezer, Orange Squeeze, or iPeng --- or if you play content using the Default skin, which always passes volume commands to LMS regardless of the Fixed Volume attribute setting. It's the job of LMS to decide what action to take, and not the client. Do some experimenting.

> _In my "mega" Group Player that we have been doing the previous testing with, the Living Room SBT is the player using your plugin and controlling my Denon AVR. The WiiM Pro Plus is also set to Fixed 100%, but does not use your plugin and I would not want the volume to change on that._

Again, it's the job of LMS --- not the client app. It checks the Fixed volume attribute and will not perform software volume control on fixed volume players.

SamInPgh commented:
> _If I understand @michaelherger suggestion correctly.... I like the idea that generically there is a means for a player to be configured as 100% Fixed with volume changes controlling and external device whether that is managed by your plugin or through a user setting a preference. Then Material could get rid of its global setting which I have never loved for the same reason stated above. In Material, my WiiM Pro Plus appears as those the volume can be changed and I would rather it not. This approach would solve that also._

We already have that. It's the player's Volume Control attribute --- "Fixed at 100%" or variable. And it's handled by LMS. Material and all the other client apps should get rid of that global setting. Every time someone installs the Denon plugin and doesn't read the Wiki pages (which means everyone), the first thing I hear is that the volume control isn't working (or, in some cases, isn't even visible) with whatever client app they are using and I have to tell them to change the option in the client app to allow volume changes for players with fixed volume. **NOT ONCE HAVE I EVER HEARD OF A COMPLAINT FROM A USER WHO SET THAT OPTION AND THEN HAD A PROBLEM WITH ANOTHER NON-DENON PLAYER.**   Bold caps for emphasis --- not shouting. 😉

mikeysas commented:
@SamInPgh, I understand your point.  If every other client behaved as the Default skin for volume control then everything would just work regardless of your plugin.  But I think the reason why many of these other clients went down the path of hiding volume controls for Fixed 100% is because people not using your plugin like the visual cue that the player has no volume control because it is fixed. 

By the way this situation predates your plugin.  When I got my first SB2, I used the IRBlaster plugin which would send IR signals to my preamp for volume change with the volume fixed at 100% on the SB2.  It actually worked very well with Web UI or the SB2’s remote.

Despite my slight preference, I will be good with whatever approach is decided here.  I do like the idea of the group volume control working to change the volume on my Denon AVR through you plugin becasue currently I don’t use Group Volume because of that.  So thanks again for starting this PR and discussion.

SamInPgh commented:
Okay. Time out. I did a lot of research snd thinking about this issue last night and believe I have come up with a solution that will satisfy all the needs and requirements expressed thus far without adding any new player attributes or methods, while providing full backwards compatibility with the existing client apps and giving them a clear path forward to eliminate the klugy user option to process volume requests for players with fixed volume. My plan is to use the Groups plugin as the first "guinea pig" app for testing this approach, and I look forward to the cooperative testing effort that this task will require. More details to come. Hint: it invoves using the 'hasDigitalOut' attribute.

SamInPgh commented:
I am ready to proceed with a new approach to addressing this issue that involves reversing all the changes made in this PR thus far and replacing them with another 2-line change to Plugin.pm. What is the best way to proceed? Is there a way to delete this PR and start over? Or should I create another commit that reverses all the changes and adds the new change? I apologize for the confusion.

EDIT: Please disregard. I have committed new changes to back out my previous commit and implement the new methodology, which I will explain shortly.

SamInPgh committed:
Remove prefs option from previous commit

Implement new method to check for fixed-volume players who need to be notified of volume changes.

Signed-off-by: SamY <syahres@gmail.com>

michaelherger commented:
You should look into "interactive git rebase", which allows you to clean up commit history. That way you could provide a single commit with just the change you want, removing all the history in between. And then you have to "force push" the change to overwrite previous attempts.

I usually prefer NOT to do this, as I'm often interested in the history of a development. And as a reviewer it can be frustrating to have to review all over again - with a larger change. But in a case like this, where the in-betweens are much more complicated than the final result, I believe it's acceptable.

But I'm not the repository owner here 😀.

mikeysas commented:
@philippe44, I just wanted to add that I have been testing for the past 5 days both the revised changed proposed here along with the related changes to the Denon AVP Control plugin that Sam just released which work together.  The combination is working very well for me.  I was helping Sam with this and some other testing as I regularly use several large Group Players including one with 8 players that includes a Denon AVP plugin player and another Fixed at 100%.  I think he has come up with a good solution here.  Thanks.

SamInPgh commented:
@philippe44 @michaelherger 

Okay. Here's what I have done with the new release of the Denon AVR Control plugin, and how it can be used to achieve the goals outlined previously in this discussion, those goals being:

1. Add the ability for LMS client apps (and the Groups plugin) to detect whether a specific player whose 'digitalVolumeControl' attribute is set to '0' (volume fixed at 100%) is still able to (and desires to) receive volume commands despite that setting.

2. Provide backward compatability, as well as a path forward, for the numerous client apps which have implemented a global option to allow volume commands for all players with 'digitalVolumeControl' set to '0', which can lead to user confusion and poterntial problems for players who cannot or should not process volume commands.

3. Provide a template that can also be used by other plugins wanting to achieve the same goals as the Denon/Marantz AVR Control plugin regarding volume control.

4. Also provide backward compatability for previous versions of LMS, as no new attributes or methods are required.

**Background**:

When a player registers with LMS, one of the capability attributes it supplies is 'HasDigitalOut', which is just what it sounds like. In almost all cases "in the wild", that attribute is set to a value of '1', the most prominent exception being the SB Radio. During LMS registration, the value of 'HasDigitalOut' is stored as a player attribute/method called, oddly enough, 'hasDigitalOut()'. Also during registration, other persistent player attributes are populated from the player prefs, with default values being used if it is a new player. Among these player prefs is the aformentioned 'digitalVolumeControl' attribute. This attribute is 'validated' in the code to make sure that it is not set to '0' (Fixed at 100%) unless 'hasDigitalOut' is set to '1'. If 'hasDigitalOut' is '0', 'digitalVolumeControl' is assigned a value of '1' regardless of the prefs setting. According to the comments, this is done to prevent a player such as the SB Radio (or its owner) from being damaged.  This validation is also done in the code that manages the player prefs menu, but it in a different way. If 'hasDigitalOut' is '0', it doesn't even expose the "Volume Control" ('digitalVolumeControl') option, keeping it set at '1' with no way to toggle it, again for player/owner safety.

So, the result of the above is that one should never encounter a player with both 'hasDigitalOut' and 'digitalVolumeControl' set to '0', i.e. doesn't have a digital output AND has its volume fixed at 100%. Enter a plugin such as the current version of Denon/Marantz AVR Control. When its player registration callback is invoked for a player that is 1) using the plugin, and 2) has its 'digitalVolumeControl' value set to '0' (Fixed at 100%, which is nearly always the case, although not required), then among its other work, it also sets the 'hasDigitalOut' attribute to '0'. While this may seem at first glance to constitute misuse of an attribute unrelated to the issue at hand, it really is not. In fact, what the combination of these two settings is doing is telling LMS and any other plugin or client that 1) the volume is fixed at 100%, and 2) control of the volume has been assumed by a plugin or other entity which needs (or at least would like) to be notified of volume control changes for the player. So, as can be seen in the latest commit for the subject PR, a client app can now determine whether to process volume commands for an individual player with the following simple conditional expression:

`if ($sprefs->client($client)->get("digitalVolumeControl") || !$client->hasDigitalOut())`

or, in plain English:

_if 'digitalVolumeControl' is '1' (not fixed at 100%) OR 'hasDigitalOut' is '0' (fixed at 100% but under software control._

I hope that I have made a sufficient case for this approach. As Mike stated, we have both been putting the code through rigorous testing for the past week, using a variety of player types, including multiple models of WiiM players, squeezelite on a RPi, DAC32, SB Players, an SB Touch, etc. The next step would be to ask Craig Drummond to phase in support for this approach in Material Skin, initially while still supporting the global player option and ultimately using this method exclusively to support a mixture of players using the Denon AVR Control plugin and others with volume fixed at 100% but not using the plugin.

Let me know your thoughts. Thanks.

SamInPgh commented:
@philippe44  Are you still following this discussion? If so, I would like to get the PR merged and published when you have time. The two issues I opened will also need attention. Just to get you up to speed, Craig Drummond has integrated this methodology into Material Skin via a new 'has_digital_out' player status query attribute (LMS 9.x) and it works well there. My next step will be to contact other client app developers such as Kurt Aholst (Squeezer) to make them aware of it also. Let me know if you have any questions or concerns. Thanks.

SamInPgh commented:
Forget the two issues mentioned above. They are now closed. All systems "go" and ready for release.
